xwayland/xwm.c

```diff
@@ -467,20 +467,39 @@ static void read_surface_title(struct wlr_xwm *xwm,
 	wlr_signal_emit_safe(&xsurface->events.set_title, xsurface);
 }
 
+static bool check_transient_loop(struct wlr_xwayland_surface *candidate,
+		struct wlr_xwayland_surface *parent) {
+	while (parent) {
+		if (candidate == parent)
+			return true;
+
+		parent = parent->parent;
+	}
+
+	return false;
+}
+
 static void read_surface_parent(struct wlr_xwm *xwm,
 		struct wlr_xwayland_surface *xsurface,
 		xcb_get_property_reply_t *reply) {
+	struct wlr_xwayland_surface *candidate = NULL;
 	if (reply->type != XCB_ATOM_WINDOW) {
 		return;
 	}
 
 	xcb_window_t *xid = xcb_get_property_value(reply);
 	if (xid != NULL) {
-		xsurface->parent = lookup_surface(xwm, *xid);
+		candidate = lookup_surface(xwm, *xid);
 	} else {
 		xsurface->parent = NULL;
 	}
 
+	if (!check_transient_loop(xsurface, candidate)) {
```

Comment:
`check_transient_loop` is called even if `xsurface->parent` is set to NULL above, which is surprising. Can we move this in the `if (xid != NULL)` block?

Reply:
Yep, definitely.